crates/zkevm_circuits/src/bn254/ec_mul/mod.rs

```diff
@@ -0,0 +1,356 @@
+use crate::demux_log_queue::StorageLogQueue;
+use crate::ethereum_types::U256;
+use crate::fsm_input_output::circuit_inputs::INPUT_OUTPUT_COMMITMENT_LENGTH;
+use crate::fsm_input_output::*;
+use crate::storage_application::ConditionalWitnessAllocator;
+
+use super::*;
+
+use self::implementation::{
+    convert_field_element_to_uint256, convert_uint256_to_field_element,
+    width_4_windowed_multiplication,
+};
+use self::input::EcMulCircuitInstanceWitness;
+
+pub mod implementation;
+pub mod input;
+
+pub const MEMORY_QUERIES_PER_CALL: usize = 3;
+pub const NUM_MEMORY_READS_PER_CYCLE: usize = 3;
+const EXCEPTION_FLAGS_ARR_LEN: usize = 4;
+
+#[derive(Derivative, CSSelectable)]
+#[derivative(Clone, Debug)]
+pub struct EcMulPrecompileCallParams<F: SmallField> {
+    pub input_page: UInt32<F>,
+    pub input_offset: UInt32<F>,
+    pub output_page: UInt32<F>,
+    pub output_offset: UInt32<F>,
+}
+
+impl<F: SmallField> EcMulPrecompileCallParams<F> {
+    pub fn from_encoding<CS: ConstraintSystem<F>>(_cs: &mut CS, encoding: UInt256<F>) -> Self {
+        let input_offset = encoding.inner[0];
+        let output_offset = encoding.inner[2];
+        let input_page = encoding.inner[4];
+        let output_page = encoding.inner[5];
+
+        let new = Self {
+            input_page,
+            input_offset,
+            output_page,
+            output_offset,
+        };
+
+        new
+    }
+}
+
+fn ecmul_precompile_inner<F: SmallField, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    x: &mut UInt256<F>,
+    y: &mut UInt256<F>,
+    scalar: &mut UInt256<F>,
+) -> (Boolean<F>, (UInt256<F>, UInt256<F>)) {
+    let base_field_params = &Arc::new(bn254_base_field_params());
+    let scalar_field_params = &Arc::new(bn254_scalar_field_params());
+
+    // We need to check for infinity prior to potential masking coordinates.
+    let point_is_infinity = is_affine_infinity(cs, (&x, &y));
+
+    // Coordinates are masked with zero in-place if they are not in field.
+    let coordinates_are_in_field = validate_in_field(cs, &mut [x, y], base_field_params);
+
+    let x = convert_uint256_to_field_element(cs, &x, base_field_params);
+    let y = convert_uint256_to_field_element(cs, &y, base_field_params);
+
+    let point_on_curve = is_on_curve(cs, (&x, &y), base_field_params);
+    let point_is_valid = point_on_curve.or(cs, point_is_infinity);
+
+    // Mask the point with zero in case it is not on curve.
+    let zero = BN256SWProjectivePoint::zero(cs, base_field_params);
+    let unchecked_point = BN256SWProjectivePoint::from_xy_unchecked(cs, x, y);
+    let point =
+        BN256SWProjectivePoint::conditionally_select(cs, point_on_curve, &unchecked_point, &zero);
+
+    // Scalar is masked with zero in-place if it is not in field.
+    let scalar_in_field = validate_in_field(cs, &mut [scalar], scalar_field_params);
+    let scalar = convert_uint256_to_field_element(cs, &scalar, scalar_field_params);
+
+    let mut result =
```

Comment:
if the point is zero, then everything is bad. ( you don't wrap this in the result)